session/storage.go

```diff
@@ -0,0 +1,135 @@
+package session
+
+import (
+	"context"
+	"encoding/json"
+	"time"
+
+	"github.com/influxdata/influxdb/v2"
+)
+
+type Store interface {
+	Set(key, val string, expiration time.Duration) error
+	Get(key string) (string, error)
+	Delete(key string) error
+	ExpireAt(key string, expireAt time.Time) error
+}
+
+var storePrefix = "sessionsv1/"
+var storeIndex = "sessionsindexv1/"
+
+type Storage struct {
+	store Store
+}
+
+func NewStorage(s Store) *Storage {
+	return &Storage{s}
+}
+
+func (s *Storage) FindSessionByKey(ctx context.Context, key string) (*influxdb.Session, error) {
+	val, err := s.store.Get(sessionIndexKey(key))
+	if err != nil {
+		return nil, err
+	}
+
+	if val == "" {
+		return nil, &influxdb.Error{
+			Code: influxdb.ENotFound,
+			Msg:  influxdb.ErrSessionNotFound,
+		}
+	}
+
+	id, err := influxdb.IDFromString(val)
+	if err != nil {
+		return nil, err
+	}
+	return s.FindSessionByID(ctx, *id)
+}
+
+func (s *Storage) FindSessionByID(ctx context.Context, id influxdb.ID) (*influxdb.Session, error) {
+	val, err := s.store.Get(storePrefix + id.String())
+	if err != nil {
+		return nil, err
+	}
+	if val == "" {
+		return nil, &influxdb.Error{
+			Code: influxdb.ENotFound,
+			Msg:  influxdb.ErrSessionNotFound,
+		}
+	}
+
+	session := &influxdb.Session{}
+	return session, json.Unmarshal([]byte(val), session)
+}
+
+func (s *Storage) CreateSession(ctx context.Context, session *influxdb.Session) error {
+	// create session
+	sessionBytes, err := json.Marshal(session)
+	if err != nil {
+		return err
+	}
+
+	// use a minute time just so the session will expire if we fail to set the expiration later
+	sessionKey := sessionKey(session.ID)
+	if err := s.store.Set(sessionKey, string(sessionBytes), time.Minute); err != nil {
+		return err
+	}
+
+	// create index
+	indexKey := sessionIndexKey(session.Key)
+	if err := s.store.Set(indexKey, session.ID.String(), time.Minute); err != nil {
+		return err
+	}
+
+	// set expiration
+	if err := s.store.ExpireAt(sessionKey, session.ExpiresAt); err != nil {
+		return err
+	}
+	if err := s.store.ExpireAt(indexKey, session.ExpiresAt); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *Storage) RefreshSession(ctx context.Context, id influxdb.ID, expireAt time.Time) error {
+	session, err := s.FindSessionByID(ctx, id)
+	if err != nil {
+		return err
+	}
+
+	if expireAt.Before(session.ExpiresAt) {
+		// no need to recreate the session if we aren't extending the expiration
+		return nil
+	}
+
+	session.ExpiresAt = expireAt
+	return s.CreateSession(ctx, session)
+}
+
+func (s *Storage) DeleteSession(ctx context.Context, id influxdb.ID) error {
+	session, err := s.FindSessionByID(ctx, id)
+	if err != nil {
+		return err
+	}
+	if session == nil {
+		return nil
+	}
+
+	if err := s.store.Delete(sessionKey(session.ID)); err != nil {
+		return err
+	}
+
+	if err := s.store.Delete(sessionIndexKey(session.Key)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func sessionKey(id influxdb.ID) string {
```

Comment:
I think this should be named `sessionID` since session has a `Key`. 
It is confusing that the ID is the key and the key is the index key.